File: Python3/FourSum.py
class Solution:
    def fourSum(self, nums: List[int], target: int) -> List[List[int]]:
        if len(nums)<4:
            return []
            
        nums.sort()
        res = []

        for i in range(len(nums)-3):
            for j in range(i+1, len(nums)-2):
                k = j + 1
                l = len(nums) - 1
                find = target - nums[i] - nums[j]
                while k < l:
                    if nums[k] + nums[l] < find:
                        k = k + 1
                    elif nums[k] + nums[l] > find:
                        l = l - 1
                    else:
                        if [nums[i], nums[j], nums[k], nums[l]] not in res:
                            res.append([nums[i], nums[j], nums[k], nums[l]])
                        k = k + 1
                        while k < l and nums[k] == nums[k-1]:
                            k = k + 1
        
        return res

File: Golang/FourSum.go
func fourSum(nums []int, target int) [][]int {

    //corner case
    if len(nums) < 4 {
        return [][]int{}
    }
    sort.Ints(nums)
    res := make([][]int, 0)

    for i:=0; i < len(nums)-3; i++ {
        if i > 0 && nums[i] == nums[i-1] {
	    //avoid duplicate res
            continue
        }
        for j:=i+1; j < len(nums)-2;j++ {
            if j > i + 1 && nums[j] == nums[j-1] {
		//avoid duplicate res
                continue
            }
            left := j + 1
            right := len(nums) - 1
            find := target - nums[i] - nums[j]
            for left < right {
                if nums[left] + nums[right] < find {
                    left = left + 1
                } else if nums[left] + nums[right] > find {
                    right = right - 1
                } else {
                    res = append(res,[]int{nums[i], nums[j], nums[left], nums[right]})
                    left = left + 1
                    for left < right && nums[left] == nums[left-1] {
                        left = left + 1
                    }
                }
            }
        }
    }
    return res
}